src/clock.rs

```diff
@@ -0,0 +1,649 @@
+
+impl CsExt for CS {
+    fn constrain(self) -> ClockConfig<NoClockDefined, NoClockDefined> {
+        ClockConfig::new(self)
+    }
+}
+
+/// Typestate for `ClockConfig` that represents unconfigured clocks
+pub struct NoClockDefined;
+/// Typestate for `ClockConfig` that represents a configured MCLK
+pub struct MclkDefined(MclkSel);
+/// Typestate for `ClockConfig` that represents a configured SMCLK
+pub struct SmclkDefined(DIVS_A);
+
+/// MODCLK frequency
+pub const MODCLK: u32 = 25_000_000;
+/// VLOCLK frequency
+pub const VLOCLK: u32 = 9_400;
+/// LFXTCLK frequency
+pub const LFXTCLK: u32 = 32_768;
+/// HFXT min frequency
+pub const HFXTMINCLK: u32 = 1_000_000;
+/// HFXT max frequency
+pub const HFXTMAXCLK: u32 = 48_000_000;
+
+///Selectable HFXTCLK frequencies
+#[derive(Debug, Clone, Copy)]
+pub enum HfxtclkFreqSel {
+    // 1-4 MHz
+    _1_4MHz,
+    // 4-8 MHz
+    _4_8MHz,
+    // 8-16 MHz
+    _8_16MHz,
+    // 16-24 MHz
+    _16_24MHz,
+    // 24-32 MHz
+    _24_32MHz,
+    // 32-40 MHz
+    _32_40MHz,
+    // 40-48 MHz
+    _40_48MHz,
+}
+
+impl HfxtclkFreqSel {
+    #[inline(always)]
+    const fn hfxtsel(&self) -> HFXTFREQ_A {
+        match *self {
+            HfxtclkFreqSel::_1_4MHz => HFXTFREQ_A::HFXTFREQ_0,
+            HfxtclkFreqSel::_4_8MHz => HFXTFREQ_A::HFXTFREQ_1,
+            HfxtclkFreqSel::_8_16MHz => HFXTFREQ_A::HFXTFREQ_2,
+            HfxtclkFreqSel::_16_24MHz => HFXTFREQ_A::HFXTFREQ_3,
+            HfxtclkFreqSel::_24_32MHz => HFXTFREQ_A::HFXTFREQ_4,
+            HfxtclkFreqSel::_32_40MHz => HFXTFREQ_A::HFXTFREQ_5,
+            HfxtclkFreqSel::_40_48MHz => HFXTFREQ_A::HFXTFREQ_6,
+        }
+    }
+
+    /// Numerical frequency
+    #[inline]
+    pub const fn freq(&self) -> u32 {
+        match *self {
+            HfxtclkFreqSel::_1_4MHz => 4_000_000,
+            HfxtclkFreqSel::_4_8MHz => 8_000_000,
+            HfxtclkFreqSel::_8_16MHz => 16_000_000,
+            HfxtclkFreqSel::_16_24MHz => 24_000_000,
+            HfxtclkFreqSel::_24_32MHz => 32_000_000,
+            HfxtclkFreqSel::_32_40MHz => 40_000_000,
+            HfxtclkFreqSel::_40_48MHz => 48_000_000,
+        }
+    }
+}
+
+/// Selectable REFOCLK frequencies
+/// Default: 32.768 KHz
+#[derive(Debug, Clone, Copy)]
+pub enum RefoclkFreqSel {
+    // 32.738 KHz
+    _32_768,
+    // 128 KHz
+    _128_000,
+}
+
+impl RefoclkFreqSel {
+    #[inline(always)]
+    fn refofsel(&self) -> REFOFSEL_A {
+        match *self {
+            RefoclkFreqSel::_32_768 => REFOFSEL_A::REFOFSEL_0,
+            RefoclkFreqSel::_128_000 => REFOFSEL_A::REFOFSEL_1,
+        }
+    }
+
+    /// Numerical frequency
+    #[inline]
+    pub fn freq(&self) -> u32 {
+        match *self {
+            RefoclkFreqSel::_32_768 => 32_768,
+            RefoclkFreqSel::_128_000 => 128_000,
+        }
+    }
+}
+
+/// Selectable DCOCLK frequencies when using factory trim settings.
+/// Actual frequencies may be slightly higher.
+#[derive(Debug, Clone, Copy)]
+pub enum DcoclkFreqSel {
```

Comment:
Maybe rename to `DCOFrequencySelection`

Reply:
Ok.

Reply:
Maybe just `DcoFrequency`? I think the fact that this is a selection is already conveyed by the fact that this is an enum.